src/format/format/input.rs

```diff
@@ -1,34 +1,37 @@
+use std::ptr::NonNull;
+
 use crate::ffi::*;
 use crate::utils;
 
+use super::Flags;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub struct Input {
-    ptr: *mut AVInputFormat,
+    ptr: NonNull<AVInputFormat>,
 }
 
 impl Input {
-    pub unsafe fn wrap(ptr: *mut AVInputFormat) -> Self {
-        Input { ptr }
-    }
-
-    pub unsafe fn as_ptr(&self) -> *const AVInputFormat {
-        self.ptr as *const _
+    pub unsafe fn from_raw(ptr: *const AVInputFormat) -> Option<Self> {
+        NonNull::new(ptr as *mut _).map(|ptr| Self { ptr })
     }
 
-    pub unsafe fn as_mut_ptr(&mut self) -> *mut AVInputFormat {
-        self.ptr
+    pub fn as_ptr(self) -> *const AVInputFormat {
+        self.ptr.as_ptr()
     }
-}
 
-impl Input {
-    pub fn name(&self) -> &str {
+    pub fn name(self) -> &'static str {
```

Comment:
Are these actually `'static`? This gives a guarantee that it will be alive for the entire life of the program, which I'm not sure is accurate.

Reply:
Yes, they are static. If you look at [libavformat/allformats.c](https://ffmpeg.org/doxygen/4.2/allformats_8c_source.html), you can see that all functions returning an `AVInputFormat`/`AVOutputFormat` will have a `ff_thread_once(..., av_format_init_next)` (= run init once) call inside.

Looking at `av_format_init_next`, you can see that variables like `muxer_list` are used to initialize, and those in turn are defined in `libavformat/muxer_list.c` and `libavformat/demuxer_list.c`. Those are auto-generated when configuring, but are basically

```c
static const AVOutputFormat * const muxer_list[] = {
  // list of AVOutputFormats here
};
```

and so on.

The internal logic has changed a bit since 4.2, but the `static const AVOutputFormat * const muxer_list[]` has stayed the same.